logstash-core/lib/logstash/util/unicode_normalizer.rb

```diff
@@ -0,0 +1,66 @@
+# Licensed to Elasticsearch B.V. under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+module LogStash
+
+  # A class to normalize the invalid unicode data
+  class UnicodeNormalizer
+
+    include LogStash::Util::Loggable
+
+    # Tries to convert input string to UTF-8 with a standard way
+    # If input string is the force encoded invalid unicode:
+    #   applies unicode normalization
+    #   replaces invalid unicode bytes with replacement characters
+    # string_data - The String data to be normalized.
+    # Returns the normalized string data.
+    def self.normalize_string_encoding(string_data)
+      input_string = string_data.dup if string_data.frozen?
+      input_string = string_data unless string_data.frozen?
+
+      if input_string.encoding != Encoding::UTF_8
+        encoding_converter = Encoding::Converter.new(input_string.encoding, Encoding::UTF_8)
+        conversion_error, utf8_string = false, nil
+        begin
+          utf8_string = encoding_converter.convert(input_string).freeze
+        rescue => e
+          # we mostly get Encoding::UndefinedConversionError but let's do not expect surprise crashes
+          logger.trace? && logger.trace("Could not convert, #{e.inspect}")
+          conversion_error = true
+        ensure
+          # if we cannot convert with a standard way
+          # we let normalize and replace invalid unicode bytes
+          return utf8_string unless conversion_error
+        end
+      end
+
+      begin
+        # non expensive `force_encoding` operation which changes the encoding metadata
+        # otherwise unicode normalization rejects
+        input_string = input_string.force_encoding(Encoding::UTF_8)
+        # force UTF-8 encoding as data might also have invalid bytes
+        # we try to normalize first, use replacement char with `scrub` if invalid bytes found
+        input_string.unicode_normalize! # use default :NFC normalization since decompositions may result multiple characters
+      rescue => e
+        logger.trace? && logger.trace("Could not normalize to unicode, #{e.inspect}")
+        logger.trace? && logger.trace("Replacing invalid non-utf bytes with replacement char.")
+        input_string.scrub!
+      end
+      input_string
+    end
```

Comment:
The simplified form without any copy operations using `String#encode`'s ability to take a source encoding instead of assuming the one that the string is flagged as:

```suggestion
    def self.normalize_string_encoding(string_data)
        # when given BINARY-flagged string, assume it is UTF-8 so that
        # subsequent cleanup retains valid UTF-8 sequences
        source_encoding = string_data.encoding
        source_encoding = Encoding::UTF_8 if source_encoding == Encoding::BINARY

        string_data.encode(Encoding::UTF_8, source_encoding, invalid: :replace, undef: :replace).scrub
    end
```

Reply:
- Nice, TIL that `force_encoding` -> `unicode_normalize` behavior can be achieved with `String#encode` 👍 Thank you!
- One thing I am not getting is enforcing to UTF-8 when source encoding is a binary (`ASCII-8BIT`) format.
- It seems we don't need `.scrub`  since we used `invalid: :replace, undef: :replace` options. Let me know if I missed anything.

Reply:
The `String#encode` does _not_ normalize the representation of code-points _unless_ the source and destination encodings are different from each other, and then only does so as defined by the specific transcoder between those encodings. We don't need to normalize the different ways of representing a code-point because doing so unnecessarily changes bytes whose values are already _equivalent_. The normalization is _usually_ useful to make binary comparisons possible, such as when using the sequence as a unique identifier.

When the encoding is _labeled_ as `BINARY`/`ASCII-8BIT`, _any_ sequence of bytes is considered "valid", and when that is converted to UTF-8 _any single byte_ above the 7-bit ASCII plane cannot be converted `UTF-8` because there is no mapping, so each of those bytes will be replaced with the unicode replacement character. If we "guess" that a `BINARY`-flagged sequence might actually be UTF-8, then when the conversion is done all 7-bit characters will translate exactly the same, but any actually-valid multibyte UTF-8 sequences will be preserved. It is _never_ more lossy, and can be less lossy in some circumstances.

The `String#scrub` seems silly, but is necessary because `String#encode` is a no-op when the source and destination encodings are identical; a string that is _flagged_ as `UTF-8` but is invalid will remain invalid when encoded to `UTF-8`.

Reply:
Agree with all points mentioned here! Thanks so so much for explaining in detail. ❤️

Reply:
The suggestion is no anymore active (probably after my rebase and force push), I have added this change with [my recent commit](https://github.com/elastic/logstash/pull/16072/commits/bc4b5342aab8c936a232fd5ed0478b7ec0476415) and added you in the co-aouthers list. Thank you so much!